src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs

```diff
@@ -159,6 +168,59 @@ internal void AddProvider(string workerDir)
             }
         }
 
+        private List<WorkerDescriptionProfile> ReadWorkerDescriptionFromProfiles(JObject workerConfig)
+        {
+            List<WorkerDescriptionProfile> descriptionProfiles = new List<WorkerDescriptionProfile>();
+            JArray profiles = workerConfig.GetValue(RpcWorkerConstants.WorkerDescriptionProfiles) as JArray;
+            if (profiles != null)
+            {
+                try
+                {
+                    //descriptionProfiles = profiles.ToObject<List<WorkerDescriptionProfile>>();
+                    foreach (JObject item in profiles)
+                    {
+                        descriptionProfiles.Add(JsonConvert.DeserializeObject<WorkerDescriptionProfile>(item.ToString(Formatting.None), new WorkerDescriptionProfileConverter(_logger, _systemRuntimeInformation, _environment)));
+                    }
+                }
+                catch (Exception ex)
+                {
+                    throw new FormatException("Failed to parse profiles in worker config.");
+                }
+            }
+            return descriptionProfiles;
+        }
+
+        private RpcWorkerDescription LoadWorkerDescriptionFromProfiles(List<WorkerDescriptionProfile> workerDescriptionProfiles, RpcWorkerDescription defaultWorkerDescription)
+        {
+            foreach (var profile in workerDescriptionProfiles)
+            {
+                if (CheckProfileConditions(profile))
+                {
+                    RpcWorkerDescription profileDescription = profile.ProfileDescription;
+                    profileDescription.Arguments = profileDescription.Arguments?.Count > 0 ? profileDescription.Arguments : defaultWorkerDescription.Arguments;
+                    profileDescription.DefaultExecutablePath = string.IsNullOrEmpty(profileDescription.DefaultExecutablePath) ? defaultWorkerDescription.DefaultExecutablePath : profileDescription.DefaultExecutablePath;
+                    profileDescription.DefaultWorkerPath = string.IsNullOrEmpty(profileDescription.DefaultWorkerPath) ? defaultWorkerDescription.DefaultWorkerPath : profileDescription.DefaultWorkerPath;
+                    profileDescription.Extensions = profileDescription.Extensions?.Count > 0 ? profileDescription.Extensions : defaultWorkerDescription.Extensions;
+                    profileDescription.Language = string.IsNullOrEmpty(profileDescription.Language) ? defaultWorkerDescription.Language : profileDescription.Language;
+                    profileDescription.WorkerDirectory = string.IsNullOrEmpty(profileDescription.WorkerDirectory) ? defaultWorkerDescription.WorkerDirectory : profileDescription.WorkerDirectory;
+                    return profileDescription;
```

Comment:
Move this logic to the profile, exposing a method called `ApplyProfile` that takes the description. 

There, you can also introduce some helpers to reduce repetition with the null/empty and other checks.